### DISCUSSION TITLE
@click on button to fetch from api not re-rendering template x-for

### DISCUSSION THREAD

**Discussion by leonardchiu** [alpinejs/alpine#2195] [1. Help] [answered]
I am new to Alpine and while exploring Alpine with Fetch to get same data came across this error.

initial x-init is ok to render via x-for.... but when i click on the button which trigger a fetch for new data, template with x-for will not re-render. Instead it throws an error: Alpine Expression Error: Cannot read properties of undefined (reading 'after')

I am guessing it is because the template x-for is not re-rendering with the new data.

Any help greatly appreciated!

**Comment by Tim-Wils** [comment]
Show us a minimal example of your problem in a codepen, so we can help you with the problem 😄👍🏻

**Comment by leonardchiu** [comment, marked as answer]
Hi Tim-Wils,

Please put in my codepen here: https://codepen.io/leonardchiu/pen/VwzZKmQ

Appreciate any help!

**Reply by SimoTod** [reply]
Your init returns an array of posts but your getone only returns a single object.
You can see the difference clicking log, you'll need to change getOne so it sets an array containing your object.
`this.blogs = [res]`

**Reply by Tim-Wils** [reply]
@SimoTod lol, fast as always! just wanted to say the same thing. 
@leonardchiu Additionally, you may want to clean up your code a bit and change your initial blogs: {} to blogs: []

**Reply by leonardchiu** [reply]
Thanks!! Really appreciate both of you jumping in to help me out! I came from Filemaker which has an awesome community where I learn a ton and constantly trying to help out there. Heard many horror stories about other tech community but obviously not here at AlpineJS github. 

@SimoTod thanks for pointing out that x-init returned an array but getOne return an object... wrapping [] around res definitely solved the issue but i am not quite sure why.... which i will research more on. Going to google "objects vs arrays"

@Tim-Wils Thanks for the heads up for cleaning up my codes. Not sure if i know whats the actual diff between an empty variable = {} vs [] vs "" but i will try to research and learn further.

I also like to list the error thrown in the console which i couldn't find the answers when googling it so that it can get indexed and help someone out in future. 

Console error: Alpine Expression Error: Cannot read properties of undefined (reading 'after')

**Reply by Tim-Wils** [reply]
When do you get this error? I don't see it while running your codepen

**Reply by leonardchiu** [reply]
<img width="1436" alt="Screenshot 2021-10-08 at 10 48 36 PM" src="https://user-images.githubusercontent.com/23194315/136577744-31166faa-39ea-4d2c-bd99-86e12b06a32c.png">

Please see screenshot...  i remove the [] at `this.blogs = res` and click on getOne... The codepen console will log the error. It is the same error on my browser console when i preview on chrome.

**Reply by Tim-Wils** [reply]
@leonardchiu ah, you’ll have to put the res object in an array []. The x-for loops through the array with your result objects and not an object itself.

**Reply by leonardchiu** [reply]
Got it! @Tim-Wils
Strange that google does return anything related to the logged error. Anyway hoping this thread help someone with similar encounter.

Cheers!

**Reply by zaengit** [reply]
this :key is duplicate
replace with: x-for="(blog, i) in blogs" :key="'blog' + i"
